Workaround for NSDecimalNumber longLongValue issue.

mortenbekditlevsen opened the pull request:
### Discussion

The discussion of the issue can be found here:
https://github.com/firebase/firebase-ios-sdk/issues/4108

This PR only provides a workaround for a bug with [NSDecimalNumber longLongValue] on values with high precision.

It does not deal with the fact that the RTDB server appears to store a double wrapped in an NSNumber differently than the same value wrapped in an NSDecimalNumber.
As an NSDecimalNumber the higher precision is stored, but the persistence layer has no knowledge about whether we are in one case or the other.

### Testing

The added logic changes the CFNumber type of one of the existing tests. Otherwise all existing tests passed.

### API Changes

No API changes in the PR.

paulb777 commented:
@mortenbekditlevsen  Thanks for the PR!  It looks like our new Danger integration is breaking CI for external PRs.  cc: @morganchen12  We'll investigate that and get this reviewed.

morganchen12 commented:
@mortenbekditlevsen The danger issue has been fixed in master, if you merge/rebase it should go away. Sorry for the inconvenience!

schmidt-sebastian commented on Firebase/Database/Persistence/FLevelDBStorageEngine.m:
```diff
@@ -878,7 +878,17 @@ - (id)fixDoubleParsing:(id)value
     // be more in line with what the server expects, so we use that here
-    if ([value isKindOfClass:[NSNumber class]]) {
+    if ([value isKindOfClass:[NSDecimalNumber class]]) {
+        NSDecimal d = [(NSDecimalNumber*)value decimalValue];
```
What do you think of renaming "d" to "original" and "result" to "rounded"?

schmidt-sebastian replied:
Please add a comment with a link to the underlying Radar here.

schmidt-sebastian commented on Firebase/Database/Persistence/FLevelDBStorageEngine.m:
```diff
@@ -878,7 +878,17 @@ - (id)fixDoubleParsing:(id)value
     // values, including 2.47. Because we use the exact bytes for hashing on the
     // server this will lead to hash mismatches. The parser of NSNumber seems to
     // be more in line with what the server expects, so we use that here
```
Can you move this existing comment into the `else if if ([value isKindOfClass:[NSNumber class]])` block?

mortenbekditlevsen replied:
Certainly, but the essential idea of what is going on inside both the NSDecimalNumber case and the NSNumber case is the same, so does this comment not apply to both?

I am updating the PR with the moved comment, but I think that perhaps it belongs to the entire method.

mortenbekditlevsen commented:
Thank you very much for the review. I have updated the PR with requested changes and added additional comments.

Have you had a chance to consider the other issue described in: #4108?
Namely that it appears that the RTDB server can represent the 'same' value in different versions based on whether it is wrapped as an NSDecimalNumber or not.
This is all fine and good, but it seems that the persistence layer can not know which representation is desired, and this could cause hashing/sync issues - and also minor precision issues based on whether you are enabling persistence or not.
